File: graphs/bfs/909-snakes-and-ladders/SnakesAndLadders.py
class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        n = len(board)

        def get_coordinates(square):
            r, c = divmod(square - 1, n)
            row = n - 1 - r
            col = c if r % 2 == 0 else n - 1 - c
            return row, col

        visited = set()
        queue = deque([(1, 0)])  # (square, moves)

        while queue:
            square, moves = queue.popleft()
            if square == n * n:
                return moves
            for roll in range(1, 7):
                next_square = square + roll
                if next_square > n * n:
                    break
                r, c = get_coordinates(next_square)
                if board[r][c] != -1:
                    next_square = board[r][c]
                if next_square not in visited:
                    visited.add(next_square)
                    queue.append((next_square, moves + 1))
        return -1

        

File: graphs/bfs/909-snakes-and-ladders/SnakesAndLadders.java
class Solution {
    public int snakesAndLadders(int[][] board) {
        
        int n = board.length;
        boolean[] visited = new boolean[n * n + 1];
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{1, 0}); // {cell, moves}
        visited[1] = true;

        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int cell = curr[0], moves = curr[1];
            if (cell == n * n) return moves;

            for (int dice = 1; dice <= 6 && cell + dice <= n * n; dice++) {
                int next = cell + dice;
                int[] pos = getPosition(next, n);
                int r = pos[0], c = pos[1];
                if (board[r][c] != -1) next = board[r][c];
                if (!visited[next]) {
                    visited[next] = true;
                    queue.offer(new int[]{next, moves + 1});
                }
            }
        }

        return -1;
    }

    private int[] getPosition(int cell, int n) {
        int r = n - 1 - (cell - 1) / n;
        int c = (cell - 1) % n;
        if ((n - r) % 2 == 0) c = n - 1 - c;
        return new int[]{r, c};
    }
}